book_source/adv_user_guide_web/SDA_Methods.Rmd

```diff
@@ -0,0 +1,182 @@
+
+```{enkf, include=TRUE}
+       
+        K <- Pf %*% t(H) %*% solve((R + H %*% Pf %*% t(H))) ## Kalman Gain
+        
+        mu.a <- mu.f + K %*% (Y - H %*% mu.f) # Analysis mean vector
+        
+        Pa   <- (diag(ncol(X)) - K %*% H) %*% Pf # Analysis covariance matrix
+        
+```
+
+The EnKF is typically used for sequential state data assimilation, but we found that EnKF lead to filter divergence when combined with our uncertain data products. Filter divergence led us to create a generalized ensemble filter that estimates process variance.
+
+### Generalized Ensemble Filter
+The generalized ensemble filter follows generally the three steps of sequential state data assimilation. But, in the generalized ensemble filter we add a latent state vector that accounts for added process variance. Furthermore, instead of solving the analysis analytically like the EnKF, we have to estimate the mean analysis vector and covariance matrix with MCMC. 
+
+#### Mapping Ensemble Output to Tobit Space
+There are some instances when we have right or left censored variables from the model forecast. For example, a model estimating species level biomass may have several ensemble members that produce zero biomass for a given species. We are considering this case a left censored state variable that needs to be mapped to normal space using a tobit model. We do this by creating two matrices with dimensions number of ensembles by state variable. The first matrix is a matrix of indicator variables (y.ind), and the second is a matrix of censored variables (y.censored). When the indicator variable is 0 the state variable (j) for ensemble member (i) is sampled. This allows us to impute a normal distribution for each state variable that contains 'missing' forecasts or forecasts of zero. 
+
+```{tobit2space,include=TRUE}
+tobit2space.model <- nimbleCode({
+    for(i in 1:N){
+      y.censored[i,1:J] ~ dmnorm(muf[1:J], cov = pf[1:J,1:J])
+      for(j in 1:J){
+        y.ind[i,j] ~ dconstraint(y.censored[i,j] > 0)
+      }
+    }
+    
+    muf[1:J] ~ dmnorm(mean = mu_0[1:J], cov = pf[1:J,1:J])
+    
+    Sigma[1:J,1:J] <- lambda_0[1:J,1:J]/nu_0
+    pf[1:J,1:J] ~ dinvwish(S = Sigma[1:J,1:J], df = J)
+    
+  })
+```
+
+
+#### Generalized Ensemble Filter Model Description
+Below is the BUGS code for the analysis full analysis model. The forecast mean an covariance are calculated from the tobit2space model above. We use a tobit likelihood in this model because there are instances when the data may be left or right censored. Process variance is included by adding a latent model state (X) with a process precision matrix (q). We update our prior on q at each time step using our estimate of q from the previous time step. 
```

Comment:
I guess there is a typo here  with double "analysis" and in the next sentence "mean and covariance"

Reply:
also did you mean "nimble" code instead of "BUGS", but also not a big deal as they're similar

Reply:
Chris always calls the model code 'BUGS' code eventhough it's in nimble. In the nimble manual it says, "We adopted the BUGS language, and we have extended it to make it more flexible." So, I guess I'll leave it unless it's still confusing?